sympy/solvers/solvers.py

```diff
@@ -272,6 +272,10 @@ def checksol(f, symbol, sol=None, **flags):
 
     if isinstance(f, BooleanAtom):
         return bool(f)
+    elif isinstance(f, int):
+        return f == 0
+    elif isinstance(f, float):
+        return abs(f) < 1e-9
```

Comment:
It would be better just to `_sympify` the input.

Reply:
I tried that, but it is causing some weird errors that will have to be dealt with. For example, `checksol(0.0001, x, 0)` returns `None` instead of `False`.

Reply:
I'll see if I can fix this though.

Reply:
It seems that the following code is causing the problem I mentioned.
https://github.com/sympy/sympy/blob/9fcdf264f3310e82a7f1d159d69812cf2422dbd0/sympy/solvers/solvers.py#L278-L280
Fixing the issue will probably require an explicit type check for `Float`. Would that be better than checking if f is an `int` or a `float`?

Reply:
The same thing happens with `Integer`: it's not just float but also `pi`, `pi+1` etc. Think about the logic here: can you list all of the possible objects that you need to check for and call isinstance for all of them?

Maybe None is a reasonable return value here. Thinking about it the empty equation does not imply that`x=3` or anything else so maybe `checksol(1, x, 0)` should return `None` to indicate that `x=3` is not something that can be inferred from the equation. I'm not sure what the precise expected semantics of `checksol` should be but we shouldn't change anything without clarifying that first.

Reply:
`Integer` incidentally gets caught by a previous condition and returns `True` if the equation is zero. So something like `checksol(Integer(0), x, 1)` returns `True`. Currently, the problem is that even something like `checksol(0.0, {})` will return `None` if `Float` is not checked for. So that needs to be fixed at least.
I would argue that, regardless of what symbols are passed in the solution, `checksol` should return `True` if the given solution satisfies all the equations in the sense that all equations can be simplified to 0 if the solution passed is substituted in and should return `False` if at least one of the equations is definitely non-zero after the solution is substituted (for example something like `checksol(pi+1, x, 0)`. If that's the case, that entire `if` condition can even be removed entirely. I've tried removing that `if` block and all tests in the solvers module seem to pass so it does not look like that behavior is being tested anywhere. 
Maybe someone else with a better idea of how `checksol` should work can help here? I'm not sure who that would be.
In any case, we could maybe push in the change where `f` is sympified so that passing an `int` or a `float` at least stops producing an `AttributeError` and I could open another PR for potentially changing the behavior for equations that have no free symbols in common with the solution keys.

Reply:
> maybe checksol(1, x, 0) should return None to indicate that x=3 is not something that can be inferred from the equation.

Also, I think this logic could cause some inconsistency issues with stuff like `checksol(x - y, {x: 0, y: 0})` (which currently returns `True` even though it is impossible to infer that `x=0` and `y=0` from `x-y=0`) or even `checksol(x, {x: 0, y: 0)` (which also returns `True` even though `y=0` cannot be inferred from the given equation. I think making `checksol` return output based on what can be definitely inferred from the equation would likely require major changes and would also change a lot of existing behavior. Any thoughts on this?

Reply:
```python
if iterable(f):
    ...
f = _sympify(f)
if not f.free_symbols:
    return f.is_zero
...
```

> Maybe someone else with a better idea of how checksol should work can help here? I'm not sure who that would be.

Look at the git blame and see who has worked on this routine.